File: internal/cloud/portal/publish_test.go
package portal

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"testing"
	"time"

	"github.com/JoseAFlores777/ccp/internal/cloud/api"
	"github.com/JoseAFlores777/ccp/internal/cloud/crypt"
	"github.com/JoseAFlores777/ccp/internal/snapshot"
)

type publishOut struct {
	Commit    api.SnapshotIn    `json:"commit"`
	Revisions []api.RevisionIn  `json:"revisions"`
	Blobs     map[string][]byte `json:"blobs"`
}

// Lo que de verdad hay que probar de «Aplicar a…» no es que el portal llame al
// API: es que la MÁQUINA pueda guardar lo que el navegador fabricó. Así que
// aquí el portal publica contra un API de mentira y Go hace luego el papel del
// agente: verifica la firma, abre el manifiesto y lo guarda en un almacén de
// verdad, que es quien recalcula el id y dice que no si no corresponde.
func TestLoQuePublicaElPortalLoPuedeGuardarUnaMaquina(t *testing.T) {
	node, err := exec.LookPath("node")
	if err != nil {
		t.Skip("sin node: no se puede ejecutar el modelo del portal")
	}
	v := buildSnapVectors(t)
	acct, err := crypt.NewAccount(v.AK)
	if err != nil {
		t.Fatal(err)
	}
	base := v.Manifest.(map[string]any)
	known := []string{}
	for _, it := range base["items"].([]any) {
		known = append(known, acct.BlobID(it.(map[string]any)["hash"].(string)))
	}
	const devA, devB = "5f1c0e2a-0000-4000-8000-00000000aaaa", "5f1c0e2a-0000-4000-8000-00000000bbbb"
	const revA1 = "cc" + "00000000000000000000000000000000000000000000000000000000000000"
	const revA2 = "dd" + "00000000000000000000000000000000000000000000000000000000000000"
	entrada := map[string]any{
		"ak": v.AK, "manifest": base, "cloud_id": v.CloudID, "known": known,
		"now":   time.Date(2026, 9, 22, 8, 0, 0, 0, time.UTC).Format(time.RFC3339),
		"edits": map[string]string{"ccp/profiles/work/overlay/CLAUDE.md": "# editado desde el portal\n"},
		"devices": []map[string]string{
			{"id": devA, "name": "mac-a"},
			{"id": devB, "name": "mac-b"},
		},
		// mac-a ya tiene dos órdenes encadenadas; mac-b ninguna. Cada cadena es
		// suya. El listado del servidor sale por `created DESC`, y `created` lo
		// pone quien publica: con un reloj desajustado la más «reciente» por
		// fecha (r1) es un eslabón ya superado. La cabeza es r2, la que nadie
		// encadena, y publicar sobre r1 sería un 409 del que no se sale.
		"chains": map[string][]map[string]string{devA: {
			{"id": revA1, "prev": ""},
			{"id": revA2, "prev": revA1},
		}},
		"heads": map[string]string{devA: revA2},
	}
	dir := t.TempDir()
	in, out := filepath.Join(dir, "in.json"), filepath.Join(dir, "out.json")
	b, err := json.Marshal(entrada)
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(in, b, 0o600); err != nil {
		t.Fatal(err)
	}
	log, err := exec.Command(node, "publish_test.mjs", in, out).CombinedOutput()
	t.Logf("%s", log)
	if err != nil {
		t.Fatalf("el portal no publicó bien: %v", err)
	}

	var got publishOut
	raw, err := os.ReadFile(out)
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(raw, &got); err != nil {
		t.Fatal(err)
	}
	comprobarSnapshot(t, acct, got)
	comprobarRevisiones(t, acct, got)
}

// comprobarSnapshot hace lo que hace client.Pull y luego el almacén local:
// verifica, abre, y GUARDA. Ese guardado es el que recalcula el id, y es donde
// se nota cualquier desviación del JSON del navegador.
func comprobarSnapshot(t *testing.T, acct *crypt.Account, got publishOut) {
	t.Helper()
	c := got.Commit
	if err := acct.Verify(c.ID, c.Parent, c.Manifest, c.Sig); err != nil {
		t.Fatalf("la firma del snapshot del portal no verifica: %v", err)
	}
	js, err := acct.OpenManifest(c.ID, c.Manifest)
	if err != nil {
		t.Fatalf("el manifiesto del portal no abre: %v", err)
	}
	var m snapshot.Manifest
	if err := json.Unmarshal(js, &m); err != nil {
		t.Fatal(err)
	}
	if acct.SnapshotID(m.ID) != c.ID {
		t.Fatal("el manifiesto no corresponde a su id en la nube")
	}
	st, err := snapshot.Open(t.TempDir())
	if err != nil {
		t.Fatal(err)
	}
	// SaveManifest recalcula el id: si el JSON del navegador se desviara un
	// byte del de Go, aquí es donde la máquina diría que no.
	if err := st.SaveManifest(&m); err != nil {
		t.Fatalf("una máquina no podría guardar lo que publicó el portal: %v", err)
	}
	if m.Trigger != "portal" {
		t.Fatalf("el snapshot tenía que decir de dónde sale, y dijo %q", m.Trigger)
	}
	// Y el blob editado: abrirlo y comprobar que su hash es el del manifiesto,
	// que es justo lo que hace el agente antes de escribir nada.
	var editado snapshot.Item
	for _, it := range m.Items {
		if it.LPath == "ccp/profiles/work/overlay/CLAUDE.md" {
			editado = it
		}
	}
	sealed, ok := got.Blobs[acct.BlobID(editado.Hash)]
	if !ok {
		t.Fatalf("el portal no subió el blob de lo que editó (%s)", editado.Hash)
	}
	data, err := acct.OpenBlob(acct.BlobID(editado.Hash), sealed)
	if err != nil {
		t.Fatalf("el blob del portal no abre: %v", err)
	}
	if string(data) != "# editado desde el portal\n" || snapshot.Hash(data) != editado.Hash ||
		editado.Size != int64(len(data)) {
		t.Fatalf("el blob no es lo editado: %q / %d", data, editado.Size)
	}
}

// comprobarRevisiones repite lo primero que hace el agente: si la firma no
// verifica con la clave de esta cuenta, eso no es una orden del dueño.
func comprobarRevisiones(t *testing.T, acct *crypt.Account, got publishOut) {
	t.Helper()
	if len(got.Revisions) != 2 {
		t.Fatalf("esperaba una revisión por equipo, salieron %d", len(got.Revisions))
	}
	for _, r := range got.Revisions {
		parts := crypt.RevisionParts{ID: r.ID, Prev: r.Prev, Device: r.DeviceID,
			Snapshot: r.Snapshot, Base: r.Base, Body: r.Body}
		if err := acct.VerifyRevision(parts, r.Sig); err != nil {
			t.Fatalf("la revisión para %s no verifica: %v", r.DeviceID, err)
		}
		if r.Snapshot != got.Commit.ID {
			t.Fatalf("la revisión no nombra el snapshot publicado: %s", r.Snapshot)
		}
		if !api.ValidID(r.ID) || r.Created.IsZero() {
			t.Fatalf("revisión mal formada: %+v", r)
		}
	}
	// El destinatario va DENTRO de la firma: cambiarlo la rompe, que es lo que
	// impide que el servidor le sirva a una máquina la orden escrita para otra.
	r := got.Revisions[0]
	parts := crypt.RevisionParts{ID: r.ID, Prev: r.Prev, Device: got.Revisions[1].DeviceID,
		Snapshot: r.Snapshot, Base: r.Base, Body: r.Body}
	if err := acct.VerifyRevision(parts, r.Sig); err == nil {
		t.Fatal("una orden desviada a otro equipo seguía verificando")
	}
}

File: internal/cloud/portal/publish_test.mjs
// Publica desde el portal con un API de mentira y deja el resultado en un
// archivo para que Go compruebe lo que cuenta: que la máquina puede GUARDAR lo
// que el navegador fabricó. El API va inyectado, así que aquí no hace falta
// servidor, ni navegador, ni red.
import { readFileSync, writeFileSync } from 'node:fs';
import * as c from './web/js/crypto.js';
import { publish, buildEdited, restore, mismoConjunto } from './web/js/publish.js';

const v = JSON.parse(readFileSync(process.argv[2], 'utf8'));
let fallos = 0;
const check = (nombre, ok, extra) => {
  if (!ok) { fallos++; console.error('FALLA', nombre, extra ?? ''); } else { console.log('ok  ', nombre); }
};

const ak = c.b64(v.ak);
const keys = {
  data: await c.deriveSubkey(ak, 'ccp/v1/data'),
  ids: await c.deriveSubkey(ak, 'ccp/v1/ids'),
  sign: await c.deriveSubkey(ak, 'ccp/v1/sign'),
};

const subidos = {};
const api = {
  putBlob: async (id, sealed) => { subidos[id] = c.b64e(sealed); },
  // `exists` sale del REGISTRO del servidor, no del bucket: un blob recién
  // subido con PUT /v1/blobs aún no está registrado, y es justo lo que el
  // portal no puede dar por perdido.
  presign: async (_op, ids) => ids.map((id) => ({ id, exists: v.known.includes(id) })),
  commitSnapshot: async (in_) => { commit = in_; return { id: in_.id }; },
  // El servidor ordena por `created DESC`, y `created` lo pone quien publica:
  // con un reloj desajustado la primera fila puede ser un eslabón ya superado.
  // Por eso el vector las sirve en ese orden torcido a propósito.
  revisions: async (dev) => (v.chains[dev] || []).map((r) => ({ id: r.id, prev: r.prev || '' })),
  publishRevision: async (r) => { revisiones.push(r); return r; },
};
let commit = null;
const revisiones = [];

const edits = new Map();
for (const [lpath, texto] of Object.entries(v.edits)) edits.set(lpath, c.utf8(texto));

const pasos = [];
const out = await publish({ api, keys }, {
  base: v.manifest, baseCloud: v.cloud_id, edits,
  devices: v.devices, now: new Date(v.now), onStep: (t) => pasos.push(t),
});

check('sube solo lo editado', out.uploaded === edits.size, out.uploaded);
check('lo recién subido cuenta como que está', out.missing.length === 0, out.missing.join(', '));
check('el commit nombra todos los blobs, incluido el nuevo',
  commit.blobs.length === v.manifest.items.length, commit.blobs.length);
check('una revisión por equipo', out.results.length === v.devices.length && out.results.every((r) => r.ok),
  JSON.stringify(out.results));
check('el padre es el snapshot editado', out.manifest.parent === v.manifest.id, out.manifest.parent);
check('lo no editado conserva su hash',
  out.manifest.items.every((it) => v.edits[it.lpath] !== undefined ||
    v.manifest.items.find((o) => o.lpath === it.lpath).hash === it.hash));
check('va contando por dónde va', pasos.length > 0);

// La revisión de cada equipo encadena sobre SU cabeza, no sobre la de al lado.
for (const r of revisiones) {
  check('encadena sobre la cabeza de ' + r.device_id, (r.prev || '') === (v.heads[r.device_id] || ''), r.prev);
}
check('todas nombran el mismo snapshot', revisiones.every((r) => r.snapshot === out.snapshot));
check('todas llevan la base, que es lo que la hace una edición y no un restore',
  revisiones.every((r) => r.base === v.cloud_id));

// buildEdited es puro: dos veces con la misma fecha da el mismo manifiesto.
const a = await buildEdited(v.manifest, edits, { now: new Date(v.now) });
const b = await buildEdited(v.manifest, edits, { now: new Date(v.now) });
check('buildEdited es determinista', JSON.stringify(a) === JSON.stringify(b));

writeFileSync(process.argv[3], JSON.stringify({ commit, revisions: revisiones, blobs: subidos }));

// Publicar sin un solo equipo no es publicar: si el diálogo deja pasar una
// lista vacía (el equipo de origen revocado, o el propio portal), lo que salía
// era un snapshot nuevo en la nube, cero órdenes y una pantalla diciendo
// «Publicado». Aquí tiene que fallar ANTES de tocar nada.
{
  const tocado = [];
  const vacio = {
    putBlob: async (id) => { tocado.push('putBlob ' + id); },
    presign: async (_op, ids) => ids.map((id) => ({ id, exists: true })),
    commitSnapshot: async (in_) => { tocado.push('commitSnapshot'); return { id: in_.id }; },
    revisions: async () => [],
    publishRevision: async (r) => { tocado.push('publishRevision'); return r; },
  };
  let err = null;
  try {
    await publish({ api: vacio, keys }, {
      base: v.manifest, baseCloud: v.cloud_id, edits, devices: [], now: new Date(v.now),
    });
  } catch (e) { err = e; }
  check('publicar sin equipos falla', err !== null, 'no lanzó');
  check('y no deja nada hecho en la nube', tocado.length === 0, tocado.join(', '));
}
// «Restaurar en <máquina>» (§10.3.1, camino 2). Lo que lo distingue de una
// edición es la base vacía: con base, la máquina reconcilia y lo suyo se
// queda; sin ella, la orden es «llega a este snapshot». Y no sube ni crea
// nada: el snapshot que se restaura ya está en la nube.
{
  const tocado = [];
  const puestas = [];
  const fake = {
    putBlob: async (id) => { tocado.push('putBlob ' + id); },
    presign: async (_op, ids) => { tocado.push('presign'); return ids.map((id) => ({ id, exists: true })); },
    commitSnapshot: async (in_) => { tocado.push('commitSnapshot'); return { id: in_.id }; },
    revisions: async (dev) => (v.chains[dev] || []).map((r) => ({ id: r.id, prev: r.prev || '' })),
    publishRevision: async (r) => { puestas.push(r); return r; },
  };
  const out = await restore({ api: fake, keys }, { snapshot: v.cloud_id, devices: v.devices });
  check('restaurar no toca el almacén de la nube', tocado.length === 0, tocado.join(', '));
  check('una orden por equipo', out.results.length === v.devices.length && out.results.every((r) => r.ok));
  check('todas nombran el snapshot que se restaura', puestas.every((r) => r.snapshot === v.cloud_id));
  check('y van SIN base: eso es lo que las hace una restauración',
    puestas.every((r) => (r.base || '') === ''), JSON.stringify(puestas.map((r) => r.base)));
  check('encadenan sobre la cabeza de cada equipo',
    puestas.every((r) => (r.prev || '') === (v.heads[r.device_id] || '')));
  // La firma tiene que verificar con base vacía: Go la comprueba aparte, aquí
  // basta con que no se cuele una firma sobre otra cosa.
  check('cada orden lleva su firma', puestas.every((r) => typeof r.sig === 'string' && r.sig.length > 0));

  let err = null;
  try { await restore({ api: fake, keys }, { snapshot: v.cloud_id, devices: [] }); } catch (e) { err = e; }
  check('restaurar sin equipos falla', err !== null, 'no lanzó');
  err = null;
  try { await restore({ api: fake, keys }, { snapshot: '', devices: v.devices }); } catch (e) { err = e; }
  check('restaurar sin snapshot falla', err !== null, 'no lanzó');

  writeFileSync(process.argv[3] + '.restore', JSON.stringify({ revisions: puestas }));
}
// Aplicar a un GRUPO (F4-1). La etiqueta viaja con cada orden y va FUERA de la
// firma: lo firmado ata la orden a SU máquina, que es lo que impide desviarla.
// Si entrara en la firma, cambiar quién está en el grupo —algo que el servidor
// puede hacer, porque la membresía no va firmada— invalidaría órdenes ya
// puestas, o peor, cambiaría a quién obedece una orden ya firmada.
{
  const puestas = [];
  const fake = {
    putBlob: async () => {},
    presign: async (_op, ids) => ids.map((id) => ({ id, exists: true })),
    commitSnapshot: async (in_) => ({ id: in_.id }),
    revisions: async (dev) => (v.chains[dev] || []).map((r) => ({ id: r.id, prev: r.prev || '' })),
    publishRevision: async (r) => { puestas.push(r); return r; },
  };
  const gid = '9f1c0e2a-0000-4000-8000-00000000cccc';
  await restore({ api: fake, keys }, { snapshot: v.cloud_id, devices: v.devices, group: gid });
  check('todas las órdenes del grupo llevan su etiqueta',
    puestas.length === v.devices.length && puestas.every((r) => r.group === gid),
    JSON.stringify(puestas.map((r) => r.group)));
  // Las firmas son las MISMAS que sin grupo: si la etiqueta entrara en la
  // firma, esto cambiaría y una máquina no podría verificar una orden cuyo
  // grupo se editó después.
  puestas.length = 0;
  await restore({ api: fake, keys }, { snapshot: v.cloud_id, devices: v.devices });
  const sinGrupo = puestas.map((r) => r.sig);
  check('sin grupo la etiqueta va vacía, no ausente', puestas.every((r) => r.group === ''));
  check('firmar no depende del grupo', sinGrupo.every((sig) => typeof sig === 'string' && sig.length > 0));
}

// mismoConjunto es lo que decide si la orden sigue siendo «la del grupo»: en
// cuanto se marca o desmarca un equipo, la etiqueta se cae.
check('mismoConjunto ignora el orden', mismoConjunto(['a', 'b'], ['b', 'a']));
check('mismoConjunto no perdona un extra', !mismoConjunto(['a'], ['a', 'b']));
check('mismoConjunto no perdona un ausente', !mismoConjunto(['a', 'b'], ['a']));

process.exit(fallos === 0 ? 0 : 1);
